### PR TITLE
feat(http): add support for listening on a unix socket

### PR DESCRIPTION
**Pull request by smoser**
This adds support for listening on a local unix socket by specifying in config http.address of "unix:/path/to/name.socket".

You can then interact with the api over that socket with:

curl --unix-socket /tmp/zot.socket http://ignored.example.com/v2/name/tags/list

**What type of PR is this?**
feature - enable listening on a unix socket.

**Which issue does this PR fix**:
N/A

**What does this PR do / Why do we need it**:
I'm looking at a use case where I populate zot storage by running a zot and loading content to it.  In this scenario there is no need for networking or exposing the zot to even localhost network.

**If an issue # is not available please add repro steps and logs showing the issue**:

**Testing done on this change**:

**Automation added to e2e**:

**Will this break upgrades or downgrades?**

**Does this PR introduce any user-facing change?**:

```release-note

```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/api/controller.go [hunk 1]
```diff
@@ -166,12 +166,23 @@ func (c *Controller) Run(reloadCtx context.Context) error {
...
-	listener, err := net.Listen("tcp", addr)
-	if err != nil {
-		return err
-	}
-
-	if c.Config.HTTP.Port == "0" || c.Config.HTTP.Port == "" {
+	var listener net.Listener
+	var err error
+
+	if strings.HasPrefix(c.Config.HTTP.Address, "unix:") {
```
Comment (andaaron) [inline pkg/api/controller.go:172]: There should be a test to cover this type of configuration.

Comment (rchincha) [reply pkg/api/controller.go:172]: Wait, why not just listen on localhost?

Comment (smoser) [reply pkg/api/controller.go:172]: > Wait, why not just listen on localhost?

Any process on the system can reach a 'localhost' endpoint while unix sockets are protected by filesystem permissions. I agree that this isn't a *huge* win, as I could protect access to the zot via httpasswd auth and a randomly generated username/key, but that comes for free with a unix socket.

I'll accept that unix socket support isn't needed at the moment, but rejecting the use of unix sockets due to the availability of 'localhost' isn't an idea that would stand on its own very long. I really did this as a proof of concept.

Comment (hallyn) [reply pkg/api/controller.go:172]: And in addition to protecting unix socket with file perms, this also opens the door for potential zot configs saying "only uid 100000-165536 can connect", using SO_PEERCRED.

**Comment by smoser** [comment]
I think this would be useful, but its not something I'm going to follow up on now.
Thanks everyone for work on zot.

**Pull request closed without merging**